### PR TITLE
feat: synchronize oidc user roles

### LINKED ISSUE #7470: Automatically sync roles from identity providers to Coder roles
A prospect would like to completely automate user additions to Coder.

Coder already automatically creates the user and syncs groups.

This 3rd component would automatically associate 1:N site-wide roles in Coder to a user. e.g., Template Admin, User Admin, Auditor, etc.

Part 2 would be automating group roles. Use and Admin

### PR DESCRIPTION
**Pull request by Emyrk**
Closes https://github.com/coder/coder/issues/7470

# What this does

Pulls the user's roles from the OIDC claims. Works almost identically to the group sync.

Role mapping is 1:many. So 1 role in OIDC can map to multiple coder roles.

```
--oidc-user-role-mapping='{"TemplateAuthors":["template-admin","user-admin"]}'
```

A new entitlement was added to control this feature

# Behaviors

- If you log in, but the `roles` claim is missing **you will be demoted to a member**. This is because there is no way to tell between an empty list claim and a missing claim. So turning on this feature with a bad OIDC config could revert all members on the site to `members`.
- If you try to manually update an OIDC user's roles with this turned on, you will get an error.

## Edit roles

You cannot manually assign roles to OIDC users since it is now controlled by this user role sync.

Screenshot from 2023-07-19 14-27-31

## Users page

(@BrunoQuaresma might want to style this later) 

Screenshot from 2023-07-19 16-06-17

Screenshot from 2023-07-19 16-06-37

## Log message if extra roles provided

```
[API] 2023-07-19 17:30:42.248 [debu]  coderd.userauth: OIDC roles ignored in assignment  user_id=5b775ca4-4d4a-4f3f-964e-356b9634f040  request_id=659d77bd-a472-4142-a9af-2ebbf6397fbf ignored=[TemplateAuthors]  assigned=[]
```

### CODE REVIEW

File: codersdk/users.go [hunk 1]
```diff
@@ -45,6 +45,7 @@ type User struct {
 	OrganizationIDs []uuid.UUID `json:"organization_ids" format:"uuid"`
 	Roles           []Role      `json:"roles"`
 	AvatarURL       string      `json:"avatar_url" format:"uri"`
+	LoginType       LoginType   `json:"login_type"`
```
Comment (Emyrk) [inline codersdk/users.go:48]: I am adding `LoginType` to the /users struct. 

This issue is being resolved soon with some open PRs I have: https://github.com/coder/coder/issues/5002

With this change, only admins can read users, so exposing this field is ok.
This solves some UI issues with this PR on not allowing role changing, and showing login type here: https://github.com/coder/coder/issues/8346

File: coderd/userauth.go [hunk 2]
```diff
@@ -684,12 +684,27 @@ type OIDCConfig struct {
 	// to groups within Coder.
 	// map[oidcGroupName]coderGroupName
 	GroupMapping map[string]string
+	// UserRoleField selects the claim field to be used as the created user's
+	// roles. If the field is the empty string, then no role updates
+	// will ever come from the OIDC provider.
+	UserRoleField string
+	// UserRoleMapping controls how groups returned by the OIDC provider get mapped
+	// to roles within Coder.
+	// map[oidcRoleName]coderRoleName
```
Comment (coadler) [inline coderd/userauth.go:693]:
```suggestion
	// map[oidcRoleName][]coderRoleName
```

File: codersdk/deployment.go [hunk 3]
```diff
@@ -1043,6 +1048,38 @@ when required by your organization's security policy.`,
...
+			Value:       &c.OIDC.UserRoleMapping,
+			Group:       &deploymentGroupOIDC,
+			YAML:        "userRoleMapping",
+		},
+		{
+			Name:        "OIDC User Role Default",
+			Description: "If user role sync is enabled, these roles are always included for all authenticated users. The 'member' role is always assigned.",
+			Flag:        "oidc-user-role-default",
+			Env:         "CODER_OIDC_USER_ROLE_DEFAULT",
+			Default:     strings.Join([]string{}, ","),
```
Comment (coadler) [inline codersdk/deployment.go:1078]: Using `strings.Join()` for the default here seems redundant. It could just be an empty string.

**Review by coadler — APPROVED** [review]

File: codersdk/deployment.go (hunk 3 above)
Comment (Emyrk) [reply codersdk/deployment.go:1078]: Fair.

**Pull request merged**